src/sources/audiosourcemp3.cpp

```diff
@@ -0,0 +1,531 @@
+
+    mad_header_finish(&madHeader);
+
+    if (MAD_ERROR_NONE != m_madStream.error) {
+        // Unreachable code for recoverable errors
+        DEBUG_ASSERT(!MAD_RECOVERABLE(m_madStream.error));
+        if (MAD_ERROR_BUFLEN != m_madStream.error) {
+            qWarning() << "Unrecoverable MP3 header error:"
+                    << mad_stream_errorstr(&m_madStream);
+            return ERR; // abort
+        }
+    }
+
+    if (m_seekFrameList.empty()) {
+        // This is not a working MP3 file.
+        qWarning() << "SSMP3: This is not a working MP3 file:"
+                << m_file.fileName();
+        return ERR; // abort
+    }
+
+    // Initialize audio stream length
+    setFrameCount(m_curFrameIndex);
+
+    // Calculate average values
+    m_avgSeekFrameCount = getFrameCount() / m_seekFrameList.size();
+    const unsigned long avgBitrate = sumBitrate / m_seekFrameList.size();
+    setBitrate(avgBitrate / 1000);
+
+    // Terminate m_seekFrameList
+    addSeekFrame(m_curFrameIndex, 0);
+
+    // Restart decoding at the beginning of the audio stream
+    m_curFrameIndex = restartDecoding(m_seekFrameList.front());
+    if (m_curFrameIndex != m_seekFrameList.front().frameIndex) {
+        qWarning() << "Failed to start decoding:" << m_file.fileName();
+        return ERR;
+    }
+
+    return OK;
+}
+
+AudioSource::diff_type AudioSourceMp3::restartDecoding(
+        const SeekFrameType& seekFrame) {
+    qDebug() << "restartDecoding @" << seekFrame.frameIndex;
+
+    if (0 == seekFrame.frameIndex) {
+        mad_frame_finish(&m_madFrame);
+        mad_synth_finish(&m_madSynth);
+    }
+    mad_stream_finish(&m_madStream);
+
+    mad_stream_init(&m_madStream);
+    mad_stream_options(&m_madStream, MAD_OPTION_IGNORECRC);
+    if (0 == seekFrame.frameIndex) {
+        mad_synth_init(&m_madSynth);
+        mad_frame_init(&m_madFrame);
+    }
+
+    // Fill input buffer
+    mad_stream_buffer(&m_madStream, seekFrame.pInputData,
+            m_fileSize - (seekFrame.pInputData - m_pFileData));
+
+    if (0 < seekFrame.frameIndex) {
+        // Muting is done here to eliminate potential pops/clicks
+        // from skipping Rob Leslie explains why here:
+        // http://www.mars.org/mailman/public/mad-dev/2001-August/000321.html
+        mad_frame_mute(&m_madFrame);
+        mad_synth_mute(&m_madSynth);
+    }
+
+    if ((0 != decodeFrameHeader(&m_madFrame.header, &m_madStream, false)) &&
+            !MAD_RECOVERABLE(m_madStream.error)) {
+        qWarning() << "Unrecoverable MP3 frame header decoding error:"
+                << mad_stream_errorstr(&m_madStream);
+        // failure
+        return getFrameCount();
+    }
+
+    return seekFrame.frameIndex;
+}
+
+void AudioSourceMp3::close() {
+    mad_synth_finish(&m_madSynth);
+    mad_frame_finish(&m_madFrame);
+    mad_stream_finish(&m_madStream);
+    m_madSynthCount = 0;
+
+    m_seekFrameList.clear();
+    m_avgSeekFrameCount = 0;
+    m_curFrameIndex = kFrameIndexMin;
```

Comment:
OK, we are an iteration ahead since we have discussed it last time. For me the general structure looks already robust only the names are a bit week, probably a result from history. 
What about this? :
Merge the close function to the destructor. 
Reintroduce the name postConstruct() for open or something else as a hint what it actually does like initalize().